### PR TITLE
Providing the way to express wasm(time) types/signatures in Rust code

### PR DESCRIPTION
**Pull request by yurydelendik**
The problem:  the have multiple ways to express signature metadata needed by wasmtime attached to a Rust function. There are several ways to define WASI interface in wasi-common and other project. This patch will try to describe and combine all macros that are needed to connect a native Rust function with its wasmtime `cranelift_codegen::ir::Signature`. (If wrapper/proxy exists, a native function arguments are expected to be converted automatically to corresponding `cranelift_codegen::ir::Type`)

TODOs:
- [x] convert WASI signature/syscall as example

### CODE REVIEW

**Comment by yurydelendik** [comment]
Docs:

```
#[wasmtime_method(module(func1_desc))]
fn func1(a: u8) -> u8 { a + 1 }
```

will generate/create a function with wasmtime-compatible ABI that looks like, and converts parameters to wasm-friendly types:

```
extern "C" fn func1(vmctx: *mut VMContext, arg1: u32) -> u32 {
  let c = |a:u8| -> u8 { a + 1 };
  c(arg1 as u8) as u32
}
mod func1_desc { ... }
```

The func1_desc will have `signature() -> ir::Signature` and `Wrapper` class to map function signature to the wasmtime_runtime export.

TODO:
- [ ] add macros to hide the Wrapper implementation details

Also,

```
#[wasmtime_trait(module(trait1_desc))]
trait Trait1 {
  fn func2(a: u8) -> u8;
}
```

produces:

```
trait Trait1 { .. }
mod trait1_desc {
  fn get_self(vmctx: *mut VMContext) -> dyn Trait1 { ... }
  extern "C" fn func2(vmctx: *mut VMContext, arg1: u32) -> u32 {
     ...
     get_self(vmctx).func2(arg1 as u8) as u32
  }
}

```

And similar to function the `Wrapper` struct will be generated to map InstanceHandle exports to trait.

TODO:
- [ ] macro to wrap implementation details on how to wrap `InstanceHandle`
- [ ] more docs and examples (please find "wasmtime-api/examples/hello2.rs" for now and use `cargo expand`)

**Comment by yurydelendik** [comment]
hello2.rs and hello3.rs expanded https://gist.github.com/yurydelendik/2c1bd0fa76642712203f9ece3680a104

**Comment by yurydelendik** [comment]
>     * Why function wrapper has InstanceHandle?

The [vmctx field](https://gist.github.com/yurydelendik/2c1bd0fa76642712203f9ece3680a104#file-hello2-expanded-rs-L51) is needed when `context()` is provided `#[wasmtime_method(...)]`. The context object is connected with instance (e.g. to get `global_exports`) rather than CL methods vmctx.

**Comment by yurydelendik** [comment]
added hello4.rs with annotated `impl` (the gist above has expanded code)

File: crates/bindings/macro/src/impl.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            });
+        }
+    }
+
+    let mod_item_vis = match vis {
+        Visibility::Inherited => quote! { pub(super) },
+        vis => quote! { #vis },
+    };
+    let mod_content = quote! {
+        #vis mod #mod_name {
```
Comment (steveeJ) [inline crates/bindings/macro/src/impl.rs]:
```suggestion
        #mod_item_vis mod #mod_name {
```

**Comment by steveeJ** [comment]
I filed a PR to your PR branch :smile: It adds basic support for a visibility modifier: https://github.com/yurydelendik/wasmtime/pull/1

File: crates/bindings/macro/src/impl.rs (hunk 1 above)
Comment (steveeJ) [reply crates/bindings/macro/src/impl.rs]: Solved by https://github.com/yurydelendik/wasmtime/pull/1

**Comment by yurydelendik** [comment]
The [zero shims PR](https://github.com/bytecodealliance/wasmtime/pull/839) addresses the same issue. `WasmTy` provides flexibility to access callers context. Closing this PR as a duplicate.

**Pull request closed without merging**